File: 13.罗马数字转整数.py
#
# @lc app=leetcode.cn id=13 lang=python3
#
# [13] 罗马数字转整数
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        mapping = {"I":1, "V":5, "X":10, "L":50, "C":100, "D":500, "M":1000}
        sum = mapping[s[0]]
        stack = s[0]
        for i in s[1:]:
            if mapping[stack] < mapping[i]:
                sum = sum - 2 * mapping[stack] + mapping[i]
                
            else:
                sum += mapping[i]
            stack = i
        
        return sum

# @lc code=end

File: 13.罗马数字转整数.c
/*
 * @lc app=leetcode.cn id=13 lang=c
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
int mapping_s(char key)
{
    int value;
    switch (key)
    {
    case 'I':
        value = 1;
        break;
    case 'V':
        value = 5;
        break;
    case 'X':
        value = 10;
        break;
    case 'L':
        value = 50;
        break;
    case 'C':
        value = 100;
        break;
    case 'D':
        value = 500;
        break;
    case 'M':
        value = 1000;
        break;
    }
    return value;
}

int romanToInt(char *s){
    int sum=0, length = strlen(s);
    char stack=s[0];
    sum += mapping_s(s[0]);

    for(int i=1; i < length; i++)
    {
        if(mapping_s(s[i]) > mapping_s(stack))
        {
            sum = sum - mapping_s(stack) * 2 + mapping_s(s[i]);
        }
        else
        {
            sum += mapping_s(s[i]);
        }
        stack = s[i];
    }
    return sum;

}
// @lc code=end

